File: routes/menu_routes.py
"""
Routes cho thực đơn và dinh dưỡng — Multi-tenant
"""
from flask import Blueprint, jsonify, request
from flask_login import current_user, login_required
from database import db, ThucDon, MonAn, get_user_don_vi_id
from datetime import date, datetime, timedelta

menu_bp = Blueprint('menu', __name__)


@menu_bp.route('/api/menu/today')
@login_required
def get_today_menu():
    """Lấy thực đơn hôm nay (theo đơn vị)"""
    today = date.today()
    dvid = get_user_don_vi_id()
    menus = ThucDon.query.filter_by(ngay=today, don_vi_id=dvid).all()
    result = {}
    for m in menus:
        result[m.bua] = m.to_dict()
    return jsonify(result)


@menu_bp.route('/api/menu/week')
@login_required
def get_week_menu():
    """Lấy thực đơn cả tuần (theo đơn vị)"""
    today = date.today()
    dvid = get_user_don_vi_id()
    start = today - timedelta(days=today.weekday())
    end = start + timedelta(days=6)
    menus = ThucDon.query.filter(
        ThucDon.ngay >= start, ThucDon.ngay <= end,
        ThucDon.don_vi_id == dvid
    ).order_by(ThucDon.ngay).all()
    
    result = {}
    for m in menus:
        day_str = m.ngay.isoformat()
        if day_str not in result:
            result[day_str] = {}
        result[day_str][m.bua] = m.to_dict()
    return jsonify(result)


@menu_bp.route('/api/menu/date/<date_str>')
@login_required
def get_menu_by_date(date_str):
    """Lấy thực đơn theo ngày (theo đơn vị)"""
    try:
        d = datetime.strptime(date_str, '%Y-%m-%d').date()
    except ValueError:
        return jsonify({'error': 'Invalid date format'}), 400
    
    dvid = get_user_don_vi_id()
    menus = ThucDon.query.filter_by(ngay=d, don_vi_id=dvid).all()
    result = {}
    for m in menus:
        result[m.bua] = m.to_dict()
    return jsonify(result)


@menu_bp.route('/api/menu', methods=['POST'])
@login_required
def create_menu():
    """Tạo thực đơn mới (gán đơn vị)"""
    data = request.json
    dvid = get_user_don_vi_id()
    td = ThucDon(
        ngay=datetime.strptime(data['ngay'], '%Y-%m-%d').date(),
        bua=data['bua'],
        don_vi_id=dvid,
        ghi_chu=data.get('ghi_chu')
    )
    if 'mon_an_ids' in data:
        mon_list = MonAn.query.filter(MonAn.id.in_(data['mon_an_ids']), MonAn.don_vi_id == dvid).all()
        td.mon_an_list = mon_list
    db.session.add(td)
    db.session.commit()
    return jsonify(td.to_dict()), 201


@menu_bp.route('/api/dishes')
@login_required
def get_all_dishes():
    """Lấy danh sách món ăn (theo đơn vị)"""
    dvid = get_user_don_vi_id()
    loai = request.args.get('loai')
    query = MonAn.query.filter_by(don_vi_id=dvid)
    if loai:
        query = query.filter_by(loai=loai)
    return jsonify([m.to_dict() for m in query.all()])


@menu_bp.route('/api/dishes/<int:dish_id>')
@login_required
def get_dish(dish_id):
    """Lấy chi tiết món ăn"""
    dish = MonAn.query.get_or_404(dish_id)
    return jsonify(dish.to_dict())


@menu_bp.route('/api/nutrition/daily/<date_str>')
@login_required
def get_daily_nutrition(date_str):
    """Lấy tổng dinh dưỡng theo ngày (theo đơn vị)"""
    try:
        d = datetime.strptime(date_str, '%Y-%m-%d').date()
    except ValueError:
        return jsonify({'error': 'Invalid date format'}), 400
    
    dvid = get_user_don_vi_id()
    menus = ThucDon.query.filter_by(ngay=d, don_vi_id=dvid).all()
    total = {'calo': 0, 'protein': 0, 'fat': 0, 'carbs': 0, 'vitamin_a': 0, 'vitamin_c': 0, 'canxi': 0, 'sat': 0}
    by_meal = {}
    
    for m in menus:
        meal_total = {'calo': 0, 'protein': 0, 'fat': 0, 'carbs': 0}
        for mon in m.mon_an_list:
            for key in total:
                total[key] += getattr(mon, key, 0)
            for key in meal_total:
                meal_total[key] += getattr(mon, key, 0)
        by_meal[m.bua] = meal_total
    
    return jsonify({
        'ngay': d.isoformat(),
        'tong': {k: round(v, 1) for k, v in total.items()},
        'theo_bua': by_meal
    })

File: routes/menuRoutes.js
/**
 * Menu routes - Thực đơn
 */
const express = require('express');
const router = express.Router();
const { db, getUserDonViId, thucDonToDict, getMonAnForThucDon } = require('../database');

const loginRequired = (req, res, next) => {
  if (!req.session?.user) return res.status(401).json({ error: 'Chua dang nhap' });
  next();
};

router.get('/api/menu/today', loginRequired, async (req, res) => {
  try {
    const today = new Date().toISOString().split('T')[0];
    const dvid = getUserDonViId(req);
    const menus = await db.all('SELECT * FROM thuc_don WHERE ngay = ? AND don_vi_id = ?', today, dvid);
    const result = {};
    for (const m of menus) { result[m.bua] = await thucDonToDict(m); }
    res.json(result);
  } catch (e) { res.status(500).json({ error: e.message }); }
});

router.get('/api/menu/week', loginRequired, async (req, res) => {
  try {
    const today = new Date();
    const dvid = getUserDonViId(req);
    const day = today.getDay();
    const start = new Date(today); start.setDate(today.getDate() - (day === 0 ? 6 : day - 1));
    const end = new Date(start); end.setDate(start.getDate() + 6);
    const startStr = start.toISOString().split('T')[0];
    const endStr = end.toISOString().split('T')[0];
    const menus = await db.all('SELECT * FROM thuc_don WHERE ngay >= ? AND ngay <= ? AND don_vi_id = ? ORDER BY ngay', startStr, endStr, dvid);
    const result = {};
    for (const m of menus) {
      if (!result[m.ngay]) result[m.ngay] = {};
      result[m.ngay][m.bua] = await thucDonToDict(m);
    }
    res.json(result);
  } catch (e) { res.status(500).json({ error: e.message }); }
});

router.get('/api/menu/date/:dateStr', loginRequired, async (req, res) => {
  try {
    const dvid = getUserDonViId(req);
    const menus = await db.all('SELECT * FROM thuc_don WHERE ngay = ? AND don_vi_id = ?', req.params.dateStr, dvid);
    const result = {};
    for (const m of menus) { result[m.bua] = await thucDonToDict(m); }
    res.json(result);
  } catch (e) { res.status(500).json({ error: e.message }); }
});

router.post('/api/menu', loginRequired, async (req, res) => {
  try {
    const dvid = getUserDonViId(req);
    const { ngay, bua, ghi_chu, mon_an_ids } = req.body;
    const r = await db.run('INSERT INTO thuc_don (ngay, bua, don_vi_id, ghi_chu) VALUES (?, ?, ?, ?)', ngay, bua, dvid, ghi_chu || null);
    if (mon_an_ids && mon_an_ids.length) {
      const ph = mon_an_ids.map(() => '?').join(',');
      const monList = await db.all(`SELECT id FROM mon_an WHERE id IN (${ph}) AND don_vi_id = ?`, ...mon_an_ids, dvid);
      for (const m of monList) { await db.run('INSERT INTO thuc_don_mon_an (thuc_don_id, mon_an_id) VALUES (?, ?)', r.lastInsertRowid, m.id); }
    }
    const td = await db.get('SELECT * FROM thuc_don WHERE id = ?', r.lastInsertRowid);
    res.status(201).json(await thucDonToDict(td));
  } catch (e) { res.status(500).json({ error: e.message }); }
});

router.get('/api/dishes', loginRequired, async (req, res) => {
  try {
    const dvid = getUserDonViId(req);
    const loai = req.query.loai;
    let dishes;
    if (loai) {
      dishes = await db.all('SELECT * FROM mon_an WHERE don_vi_id = ? AND loai = ?', dvid, loai);
    } else {
      dishes = await db.all('SELECT * FROM mon_an WHERE don_vi_id = ?', dvid);
    }
    res.json(dishes);
  } catch (e) { res.status(500).json({ error: e.message }); }
});

router.get('/api/dishes/:dishId', loginRequired, async (req, res) => {
  try {
    const dish = await db.get('SELECT * FROM mon_an WHERE id = ?', req.params.dishId);
    if (!dish) return res.status(404).json({ error: 'Not found' });
    res.json(dish);
  } catch (e) { res.status(500).json({ error: e.message }); }
});

router.get('/api/nutrition/daily/:dateStr', loginRequired, async (req, res) => {
  try {
    const dvid = getUserDonViId(req);
    const menus = await db.all('SELECT * FROM thuc_don WHERE ngay = ? AND don_vi_id = ?', req.params.dateStr, dvid);
    const total = { calo: 0, protein: 0, fat: 0, carbs: 0, vitamin_a: 0, vitamin_c: 0, canxi: 0, sat: 0 };
    const byMeal = {};
    for (const m of menus) {
      const mealTotal = { calo: 0, protein: 0, fat: 0, carbs: 0 };
      const monList = await getMonAnForThucDon(m.id);
      monList.forEach(mon => {
        Object.keys(total).forEach(k => { total[k] += mon[k] || 0; });
        Object.keys(mealTotal).forEach(k => { mealTotal[k] += mon[k] || 0; });
      });
      byMeal[m.bua] = mealTotal;
    }
    res.json({
      ngay: req.params.dateStr,
      tong: Object.fromEntries(Object.entries(total).map(([k, v]) => [k, Math.round(v * 10) / 10])),
      theo_bua: byMeal
    });
  } catch (e) { res.status(500).json({ error: e.message }); }
});

module.exports = router;
